### PR TITLE
Add swap page

### LINKED ISSUE #43: Adapt to updated action type `ethereum-call-contract`
Update comit-i as per below
Dod:
- [x] `ethereum-invoke-contract` is now `ethereum-call-contract`
- [x] amount field is removed from this action type

### LINKED ISSUE #51: More informative swap status
All the information we currently display about the status of the swap is derived from this enum in comit-rs:

``` rust
pub enum SwapStatus {
    InProgress,
    Swapped,
    NotSwapped,
    InternalFailure,
}
```
Together with the available actions, one can infer the state of the swap, but this should be explicit.

To solve this, a new page should be added to comit-i, which contains the whole "state" of the swap as returned by the comit node. Users can navigate to the new page by clicking on the "SwapRow".

All transaction should be displayed and linked to the proper explorer (only for public chains: testnet & mainnet).

Design:
Two columns, one for alpha and one for beta.
The goal is to explicitly communicate to the user, what they are selling and what they are buying.

### LINKED ISSUE #58: Add link to protocol's RFC
There's currently a button labelled `human-protocol-spec` which does not work. It takes up too much space and is not an action.

May16::150255

I propose to let it be a hyperlink on the protocol's name instead.

### PR DESCRIPTION
**Pull request by luckysori**
Run it locally if you can to test the swap page. You can access a swap page by clicking on a swap in the list of swaps.

Resolves #51.
Resolves #43.
Resolves #58.

### CODE REVIEW

File: cypress/fixtures/listOfTwoSwaps.json [hunk 1]
```diff
@@ -1,62 +1,102 @@
...
           "alpha_ledger": {
-            "name": "Bitcoin",
+            "name": "bitcoin",
             "network": "regtest"
           },
           "beta_asset": {
-            "name": "Ether",
-            "quantity": "10000000000000000000"
+            "name": "ether",
+            "quantity": "1000000000000000000"
```
Comment (da-kami) [inline cypress/fixtures/listOfTwoSwaps.json]: 🙃 not that it should make a difference, but I wonder why the quantity changed ;-)

File: src/api/getSwap.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
+import axios from "axios";
+import apiEndpoint from "./apiEndpoint";
+
+export interface Asset {
```
Comment (thomaseizinger) [inline src/api/getSwap.ts]: :mega: Please use the types we already have.

File: src/pages/ListSwaps/SwapRow.tsx [hunk 3]
```diff
@@ -40,208 +29,52 @@ function AssetCell({ asset }: AssetCellProps) {
...
-        }
-        break;
-    }
-  };
-
   return (
-    <React.Fragment key={swap._links.self.href}>
-      <TableRow data-cy="swap-row">
-        <TableCell>{swap.parameters.alpha_ledger.name}</TableCell>
+    <React.Fragment key={swap.links[0].href}>
```
Comment (thomaseizinger) [inline src/pages/ListSwaps/SwapRow.tsx]: :mega: Please use the `self` link and not just the first one.

File: src/pages/ListSwaps/SwapRow.tsx [hunk 4]
```diff
@@ -40,208 +29,52 @@ function AssetCell({ asset }: AssetCellProps) {
...
+  swap: Swap;
 }
 
-interface LedgerParamsField {
-  name: string;
-  value: string;
-}
+function SwapRow({ swap, history }: SwapRowProps) {
+  const classes = useStyles();
+  const swapLink = swap.links.filter(link => link.rel[0] === "self")[0].href;
```
Comment (thomaseizinger) [inline src/pages/ListSwaps/SwapRow.tsx]:
```suggestion
  const swapLink = swap.links.find(link => link.rel.contains("self)).href;
```

File: src/pages/ListSwaps/SwapRow.tsx [hunk 5]
```diff
@@ -40,208 +29,52 @@ function AssetCell({ asset }: AssetCellProps) {
...
-  };
-
   return (
-    <React.Fragment key={swap._links.self.href}>
-      <TableRow data-cy="swap-row">
-        <TableCell>{swap.parameters.alpha_ledger.name}</TableCell>
+    <React.Fragment key={swap.links[0].href}>
+      <TableRow
+        hover={true}
+        onClick={() => history.push(swapLink)}
```
Comment (thomaseizinger) [inline src/pages/ListSwaps/SwapRow.tsx]: So you are using the API link to navigate to the page locally? Do we already return the ID from the comit-node? Would probably be easier to use just that (as long as you are still passing the actual `self` link to the page so that it can fetch it from the comit-node).

File: src/api/getSwap.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  rel: string[];
+  href: string;
+}
+
+export interface Swap {
+  properties: Properties;
+  links: Link[];
+}
+
+export default function getSwap(protocol: string, swapId: string) {
```
Comment (thomaseizinger) [inline src/api/getSwap.ts]: :mega: This is exactly what you should NOT be doing :)
Please use the link provided by the comit-node to fetch the swap. Clients should not make any assumptions about what the URLs of the server look like.

File: cypress/fixtures/listOfTwoSwaps.json (hunk 1 above)
Comment (luckysori) [reply cypress/fixtures/listOfTwoSwaps.json]: Haha, well, instead of manually adding the swaps I just copied the json the comit node returns after sending two swap requests <:^)

File: src/api/getSwap.ts (hunk 2 above)
Comment (luckysori) [reply src/api/getSwap.ts]: Do you mean the ones in `getSwaps.ts`? I will combine them!

File: src/pages/ListSwaps/SwapRow.tsx (hunk 3 above)
Comment (luckysori) [reply src/pages/ListSwaps/SwapRow.tsx]: Didn't notice that! I guess it's just a key, but I'll use `swapLink` instead.

File: src/pages/ListSwaps/SwapRow.tsx (hunk 4 above)
Comment (luckysori) [reply src/pages/ListSwaps/SwapRow.tsx]: This is cleaner, but `find` returns the first element that matches the predicate or undefined. Do we want to handle that case or is it safe to assume that the comit node's response will always include a `self` link?

File: src/pages/ListSwaps/SwapRow.tsx (hunk 5 above)
Comment (luckysori) [reply src/pages/ListSwaps/SwapRow.tsx]: We can only get the swap ID from the swap link. Would it be better to extract it from it? I'd still want the path of a swap page to be `swaps/<protocol>/:id`.

File: src/api/getSwap.ts (hunk 6 above)
Comment (luckysori) [reply src/api/getSwap.ts]: I see the problem now! Does this mean that I have to pass the URL of the swap as a query parameter to the URL of the swap page? That would look so ugly :sob:

File: src/api/getSwap.ts (hunk 2 above)
Comment (thomaseizinger) [reply src/api/getSwap.ts]: I think there are more generic ones somewhere in the codebase. You should rely on `quantity` for example. Not every asset is guaranteed to have that field :)

File: src/pages/ListSwaps/SwapRow.tsx (hunk 4 above)
Comment (thomaseizinger) [reply src/pages/ListSwaps/SwapRow.tsx]: Currently, you are `[0]` the result of `filter`, which might yield an array with 0 elements, so that behaviour should stay the same? :)
If we don't get a self-link, that is a bug in the comit node.

File: src/api/getSwap.ts (hunk 6 above)
Comment (thomaseizinger) [reply src/api/getSwap.ts]: Good point, I am not quite sure how to solve this one.
The problem is deep-linking, e.g. when someone takes the URL from the `Swap` page and completely reloads it. The only state you get is the URL so you will need to be able to restore the page state from that.

@LLFourn had an interesting idea that might work:

You have one general, client-side route: `/show_resource` (name subject to be change). You append any server-link onto that:

- `/show_resource/swaps`
- `/show_resource/swaps/rfc003/18047108476127401`

Based on the response, the client will decide, which content to render (check the `class` property of the root-entity returned from the comit-node. it will be `swaps` for the list and `swap` for a single one).

Comment (luckysori) [reply src/api/getSwap.ts]: I think I get it. Would it be bad to use `/` as said general, client-side route. Otherwise we'd have to:
- Use a different home page to list swaps.
- Re-route (?) from `/` to `/swap_resource/swaps`.

File: src/pages/ListSwaps/SwapRow.tsx (hunk 4 above)
Comment (luckysori) [reply src/pages/ListSwaps/SwapRow.tsx]: I agree. The thing is the compiler complains about the result of `swap.links.find(link => link.rel.includes("self)` being possibly undefined, whereas before it didn't because it can't reason about array length.

File: src/api/getSwap.ts (hunk 2 above)
Comment (luckysori) [reply src/api/getSwap.ts]: > I think there are more generic ones somewhere in the codebase.

I think you're referring to what we do in [ledgerSpec.ts](https://github.com/comit-network/comit-i/blob/8ab8c4dff0eb38f85ec7659c50a3e6e5922663a0/src/ledgerSpec.ts), but I'm not sure how to apply that here. I don't know how to handle different shapes for the asset other than using optional fields. For now, I would just add `token_contract` as an optional field and use it if it's present and the asset name is `erc20`.

File: src/api/getSwap.ts (hunk 6 above)
Comment (luckysori) [reply src/api/getSwap.ts]: Redirecting seems best. At the moment the `class` property is empty for all resources though, so it's virtually impossible to tell them apart.

Also, this generic approach to fetching resources from the comit node means that the error handling should also be generic i.e. it's harder to display the correct error snackbar if you only know what you were fetching after you receive it.

Comment (luckysori) [reply src/api/getSwap.ts]: I've patched up the comit node locally and it works with your suggestion! Have a look at the changes when you can :smile:

Comment (thomaseizinger) [reply src/api/getSwap.ts]: Oh, that one should not be empty :)
Best to open a ticket for that so we can fix it!

Good point in regards to the error handling. Would it be sufficient if we differentiate between "failed to connect" and "bad JSON" (as suggested in #59 ?). That one should be possible even if we have generic resource loading.

File: src/pages/ListSwaps/SwapRow.tsx [hunk 7]
```diff
@@ -1,247 +1,71 @@
...
-        : swap.parameters.alpha_ledger;
-    case "refund":
-    case "fund":
-      return swap.role === "Alice"
-        ? swap.parameters.alpha_ledger
-        : swap.parameters.beta_ledger;
+  const swapLink = swap.links.filter(link => link.rel[0] === "self")[0].href;
+  function onRowClick() {
+    history.push("/show_resource" + swapLink);
+    /* FIXME: This is bad, but otherwise the Swap component doesn't mount */
```
Comment (thomaseizinger) [inline src/pages/ListSwaps/SwapRow.tsx]: This is likely a problem with the routing setup then.

File: src/pages/ListSwaps/SwapRow.tsx (hunk 4 above)
Comment (thomaseizinger) [reply src/pages/ListSwaps/SwapRow.tsx]: I think for now, we can fail, if we don't have a self link. It is not nice but the better solution (disable click handler and derive a `key` in a different, deterministic manner) is probably an overkill for now, given that it is definitely a bug in the comit node if we don't have a self link.

File: src/pages/ShowResource.tsx [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    promiseFn: getComitResourceFn,
+    resourcePath
+  });
+
+  if (isLoading) {
+    return <CenteredProgress title="Fetching..." />;
+  } else if (data && data.class.includes("swaps")) {
+    const resource = data as GetSwapsResponse;
+    return <SwapList swaps={resource.entities} />;
+  } else if (data) {
```
Comment (thomaseizinger) [inline src/pages/ShowResource.tsx]:
```suggestion
  } else if (data && data.class.includes("swap")) {
```

File: src/pages/SwapPage/BlockchainLog.tsx [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+  alphaState,
+  betaState,
+  role,
+  actions
+}: BlockchainLogProps) {
+  const alphaActions =
+    role === Role.Alice
+      ? actions.filter(
+          elem =>
+            elem.action === "deploy" ||
```
Comment (thomaseizinger) [inline src/pages/SwapPage/BlockchainLog.tsx]: :thinking: Not sure how good it is to include those here in terms of re-usability for future protocols. What was the reason why you are filtering them?

Comment (luckysori) [reply src/pages/SwapPage/BlockchainLog.tsx]: > What was the reason why you are filtering them?

In this case I wanted actions to be displayed on the right column and I believe actions don't identify what ledger they're acting on so I had to look at the role instead.

> Not sure how good it is to include those here in terms of re-usability for future protocols. 

I agree with you and a lot of our components are already too specific to RFC003, but it's very hard to anticipate how other protocols would look like. This blockchain event log component is actually just a _RFC003_ blockchain event log. Maybe the name should reflect that.

File: src/App.tsx [hunk 10]
```diff
@@ -110,7 +117,12 @@ function App({ classes, history }: AppProps) {
       </Drawer>
       <main className={classes.content}>
         <div className={classes.toolbar} />
-        <Route exact={true} path="/" component={ListSwaps} />
+        <Route
+          exact={true}
+          path="/"
+          render={() => <Redirect to="/show_resource/swaps" />}
+        />
+        <Route path="/show_resource" component={ShowResource} />
```
Comment (thomaseizinger) [inline src/App.tsx]: Not sure if this helps with your reload-problem but:

You can specify path-variables in these routes. https://reacttraining.com/react-router/web/example/url-params

File: src/pages/ListSwaps/LedgerActionDialogBody.tsx [hunk 11]
```diff
@@ -11,14 +11,19 @@ import Web3SendTransactionButton from "../../components/Web3SendTransactionButto
...
   action: LedgerAction;
 }
 
 function LedgerActionDialogBody({
   onClose,
   action
 }: LedgerActionDialogBodyProps) {
+  function onSuccess() {
+    onClose();
+    location.reload();
```
Comment (thomaseizinger) [inline src/pages/ListSwaps/LedgerActionDialogBody.tsx]: Triggering a reload in an SPA is a bit dodgy. Why did you have to add that?

File: src/pages/ShowResource.tsx [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+function ShowResource({ location }: RouteComponentProps) {
+  const resourcePath = location.pathname.replace("/show_resource/", "");
+
+  const { data, isLoading, error } = useAsync({
+    promiseFn: getComitResourceFn,
+    resourcePath,
+    /* FIXME: If connection lost whilst looking at a swap page and then
+       try to navigate back to swap list, nothing happens */
+    watch: location.pathname
```
Comment (thomaseizinger) [inline src/pages/ShowResource.tsx]: IMO, the hook should re-trigger if its arguments change. If that is not the case, then we need to drop this library :sweat_smile:

Comment (thomaseizinger) [reply src/pages/ShowResource.tsx]: But maybe this is related to the way the routing is set up?

File: src/pages/SwapPage/AssetCard.tsx [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+interface AssetCardProps {
+  title: string;
+  ledger: Ledger;
+  asset: Asset;
+  tradeAction: string;
+  expiry: number;
+}
+
+/* TODO: Handle the fact that this code is specific to RFC003 due
```
Comment (thomaseizinger) [inline src/pages/SwapPage/AssetCard.tsx]: Don't bother with this one too much for now :)

File: src/pages/SwapPage/BlockchainLog.tsx (hunk 9 above)
Comment (thomaseizinger) [reply src/pages/SwapPage/BlockchainLog.tsx]: > This blockchain event log component is actually just a RFC003 blockchain event log.

Fair enough then, this one can have knowledge about those terms! :)

File: src/pages/SwapPage/CommunicationCard.tsx [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+  status: CommunicationStatus;
+  role: Role;
+}
+
+function CommunicationCardHeader({
+  status,
+  role
+}: CommunicationCardHeaderProps) {
+  /* TODO: Display swap decline reason if available */
+  /* TODO: Add retry button */
```
Comment (thomaseizinger) [inline src/pages/SwapPage/CommunicationCard.tsx]: Some serious scope creep here? I am not aware that we have the ability to retry the sending :D

File: src/api/swapTypes.ts [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+export function toMainUnit(asset: Asset) {
+  switch (asset.name) {
+    case "ether":
+      return fromWei(asset.quantity, "ether") + " ETH";
+    case "bitcoin":
+      return toBitcoin(asset.quantity) + " BTC";
+    case "erc20":
+      // TODO: Find and use token symbol if possible
```
Comment (da-kami) [inline src/api/swapTypes.ts]: ❓ Did you create a ticket for this?

File: src/api/swapTypes.ts [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+export interface Asset {
+  name: string;
+  quantity: string;
+  token_contract?: string;
+}
+
+export function toMainUnit(asset: Asset) {
+  switch (asset.name) {
+    case "ether":
+      return fromWei(asset.quantity, "ether") + " ETH";
```
Comment (da-kami) [inline src/api/swapTypes.ts]: 🔧 The result of the conversion functions for the different assets produces different results when displayed in the UI (which is a little weird). We should probably unify the result after the conversion so it is displayed the same way.
Currently when sending 1 Satoshi to 1 Wei:
Bitcoin:
```
Selling 1e-8 BTC on bitcoin testnet
```
Ether:
```
Buying 0.000000000000000001 ETH on ethereum ropsten
```

File: src/pages/ListSwaps/LedgerActionDialogBody.tsx (hunk 11 above)
Comment (luckysori) [reply src/pages/ListSwaps/LedgerActionDialogBody.tsx]: I didn't have to, but otherwise after performing an ethereum action the button is still there, because the resource (either the swap list or a particular swap) hasn't been fetched again.

File: src/pages/ShowResource.tsx (hunk 12 above)
Comment (luckysori) [reply src/pages/ShowResource.tsx]: I need to investigate this some more, but I think the library is fine.

File: src/App.tsx (hunk 10 above)
Comment (luckysori) [reply src/App.tsx]: I'm not sure how to use that to express that a re-render should happen when the url goes from `localhost:8000/show_resource/swaps` to `localhost:8000/show_resource/swaps/rfc003/:id`.

File: src/pages/SwapPage/CommunicationCard.tsx (hunk 14 above)
Comment (luckysori) [reply src/pages/SwapPage/CommunicationCard.tsx]: Hahah, I did not mean to imply that it all should be done with this PR! I can delete the TODOs if you prefer.

File: src/api/swapTypes.ts (hunk 16 above)
Comment (luckysori) [reply src/api/swapTypes.ts]: Good point! I never noticed because we always try to test with 1BTC for 1ETH.

File: src/App.tsx (hunk 10 above)
Comment (thomaseizinger) [reply src/App.tsx]: I was more thinking of using a path variable for the resource link:

`<Route path="/show_resource/:resourceLink">`

File: src/api/swapTypes.ts (hunk 15 above)
Comment (luckysori) [reply src/api/swapTypes.ts]: No... I ended up removing almost all of these TODOs because they're out of scope for the time being.

File: src/pages/ListSwaps/LedgerActionDialogBody.tsx (hunk 11 above)
Comment (thomaseizinger) [reply src/pages/ListSwaps/LedgerActionDialogBody.tsx]: Okay, @da-kami and me are going to revise these dialogs now anyway so it's fine.

File: src/pages/ListSwaps/SwapRow.tsx [hunk 17]
```diff
@@ -1,247 +1,90 @@
...
               }}
-              color="primary"
-            >
-              {"Submit"}
-            </Button>
-            <Button
-              onClick={() => setDialogState(DialogState.Closed)}
-              color="secondary"
+              variant="outlined"
+              href={protocolSpecLink.href}
```
Comment (thomaseizinger) [inline src/pages/ListSwaps/SwapRow.tsx]: Why is this a button and not just an (external) link?

File: src/pages/ShowResource.tsx (hunk 12 above)
Comment (luckysori) [reply src/pages/ShowResource.tsx]: This was fixed, btw. The hook was triggered because of the change in the path, it failed to fetch, but the data object didn't change. The problem was with the way I ordered the if statements afterwards.
